datalad_next/iter_collections/gitworktree.py

```diff
@@ -250,23 +254,27 @@ def _lsfiles_line2props(
 
 
 def _git_ls_files(path, *args):
-    # we use a plain runner to avoid the overhead of a GitRepo instance
-    runner = ThreadedRunner(
-        cmd=[
-            'git', 'ls-files',
-            # we rely on zero-byte splitting below
-            '-z',
-            # otherwise take whatever is coming in
-            *args,
-        ],
-        protocol_class=StdOutCaptureGeneratorProtocol,
-        stdin=DEVNULL,
-        # run in the directory we want info on
-        cwd=path,
-    )
-    line_splitter = LineSplitter('\0', keep_ends=False)
-    # for each command output chunk received by the runner
-    for content in runner.run():
-        # for each zerobyte-delimited "line" in the output
-        for line in line_splitter.process(content.decode('utf-8')):
-            yield line
+    with run(
+            cmd=[
+                'git', 'ls-files',
+                # we rely on zero-byte splitting below
+                '-z',
+                # otherwise take whatever is coming in
+                *args,
+            ],
+            protocol_class=StdOutCaptureGeneratorProtocol,
+            stdin=DEVNULL,
+            cwd=path
+    ) as r:
+        # This code uses the data processor chain to process data. This fixes
+        # a problem with the previous version of the code, where `decode` was
+        # used on every data chunk that was sent tp `pipe_data_received`. But
+        # data is chunked up randomly and might be split in the middle of a
+        # character encoding, leading to weird errors.
+        yield from process_from(
+            data_source=r,
+            processors = [
+                decode_processor('utf-8'),
+                splitlines_processor(separator='\0', keep_ends=False)
+            ]
+        )
```

Comment:
```suggestion
    with run(
            cmd=[
                'git', 'ls-files',
                # we rely on zero-byte splitting below
                '-z',
                # otherwise take whatever is coming in
                *args,
            ],
            protocol_class=StdOutCaptureGeneratorProtocol,
            processors = [
                decode_processor('utf-8'),
                splitlines_processor(separator='\0', keep_ends=False)
            ],
            cwd=path,
    ) as r:
        yield from r
```

I brought it up before, but still think this usage should not be more complex than this. If `run()` must not have `processors=` it could be renamed and the replacement could have it.

Reply:
Thanks. I think I missunderstood your earlier comments.


I have a commit which defines a `StdOutCapturePipelineGeneratorProtocol`. The protocol takes an optional processor list as a keyword-argument and sends the process-pipeline results to the generator. This is transparent to the run-context-manager and achieves the same terse "ls-files"-implementation, i.e. the code in the context is just:

```
yield from r
```

I did not yet put it in because it requires selecting the "right" protocol and providing a processor list in the `protocol_kwargs`. That seemed complex to me. And, in contrast to your suggestion (put the pipeline processing into `run`) it requires support by the protocol. It would be transparent for the `run`-context manager code though.

Your suggestion should in principle support any protocol. It would require different treatment of generator- and non-generator protocols, if we want to apply pipelines to non-generator protocols as well. The latter would require the result of the non-generator protocols to adhere to a fixed convention. That seems the be the case, all non-generator protocol results are a dictionary with the keys `stdout` and `stderr`. If we want to support processor pipelines on non-generator protocols, the `run`-context manager would have to pass the values of `result_dict['stdout']` and possibly `result_dict['stderr']` through the processing pipeline. This would also require a way to specify processor lists for stdout and stderr. Alternatively, we could limit the `processors`-keyword validity to generator protocols and stdout only.

I am not sure what would be the best approach here. WDYT?

Reply:
After having worked with adding processing pipelines to non-generator protocols, I would say: "let's not do that". The non-generator protocols are very much centered around joining individual byte-chunks in their `pipe_data_received`-method and decoding them into strings in their `_prepare_result`-method.  Without overwriting the `_prepare_result`-method, it is not even possible to receive undecoded byte-content from a subprocess.

To make use of arbitrary pipelines, `pipe_data_received` and `_prepare_result` have to be overridden. That would introduce another "subtree" of protocol-class definitions. Alternatively, we could introduce a "synchronized"-run mode, that would collect all results from a  generator-protocol-based `run` and return them.

Reply:
I brought the processing protocol classes back. Using those classes the code that executes `git ls-files` looks like this:

```python
    with run(
            cmd=[
                'git', 'ls-files',
                # we rely on zero-byte splitting below
                '-z',
                # otherwise take whatever is coming in
                *args,
            ],
            protocol_class=StdOutCaptureProcessingGeneratorProtocol,
            stdin=DEVNULL,
            cwd=path,
            protocol_kwargs=dict(
                processors=[
                    decode_processor('utf-8'),
                    splitlines_processor(separator='\0', keep_ends=False)
                ]
            )
    ) as r:
        # This code uses the data processor chain to process data. This fixes
        # a problem with the previous version of the code, where `decode` was
        # used on every data chunk that was sent tp `pipe_data_received`. But
        # data is chunked up randomly and might be split in the middle of a
        # character encoding, leading to weird errors.
        yield from r
```